drivers/i2c/i2c_esp32.c

```diff
@@ -0,0 +1,781 @@
+				    int matrix_out, int matrix_in)
+{
+	const char *device_name = i2c_esp32_get_gpio_for_pin(pin);
+	struct device *gpio;
+
+	if (!device_name) {
+		return -EINVAL;
+	}
+	gpio = device_get_binding(device_name);
+	if (!gpio) {
+		return -EINVAL;
+	}
+
+	gpio_pin_configure(gpio, pin,
+			   GPIO_DIR_OUT | GPIO_DS_DISCONNECT_LOW |
+			   pull_up ? GPIO_PUD_PULL_UP : GPIO_PUD_NORMAL);
+	gpio_pin_write(gpio, pin, default_level);
+
+	gpio_matrix_out(pin, matrix_out, 0, 0);
+	gpio_matrix_in(pin, matrix_in, 0);
+
+	return 0;
+}
+
+static void i2c_esp32_configure_frequency(const struct i2c_esp32_config *config,
+					  u32_t freq_hz)
+{
+	const u32_t period = (APB_CLK_FREQ / freq_hz) / 2;
+
+	*config->regs.scl_low_period = period << I2C_SCL_LOW_PERIOD_S;
+	*config->regs.scl_high_period = period << I2C_SCL_HIGH_PERIOD_S;
+
+	*config->regs.scl_start_h_time =
+		(period / 2) << I2C_SCL_START_HOLD_TIME_S;
+	*config->regs.scl_rstart_s_time =
+		(period / 2) << I2C_SCL_RSTART_SETUP_TIME_S;
+	*config->regs.scl_stop_h_time =
+		(period / 2) << I2C_SCL_STOP_HOLD_TIME_S;
+	*config->regs.scl_stop_s_time =
+		(period / 2) << I2C_SCL_STOP_SETUP_TIME_S;
+
+	*config->regs.sda_hold_time = (period / 4) << I2C_SDA_HOLD_TIME_S;
+	*config->regs.sda_sample_time = (period / 4) << I2C_SDA_SAMPLE_TIME_S;
+}
+
+static int i2c_esp32_configure(struct device *dev, u32_t dev_config_raw)
+{
+	const struct i2c_esp32_config *config = dev->config->config_info;
+	struct i2c_esp32_data *data = dev->driver_data;
+	union dev_config dev_config = { .raw = dev_config_raw };
+	int key = irq_lock();
+	u32_t v;
+	int r;
+
+	r = i2c_esp32_configure_pins(config->pins.scl,
+				     config->pins.scl_pull_up,
+				     config->pins.default_level,
+				     config->sig.scl_out,
+				     config->sig.scl_in);
+	if (r < 0) {
+		return r;
+	}
+
+	r = i2c_esp32_configure_pins(config->pins.sda,
+				     config->pins.sda_pull_up,
+				     config->pins.default_level,
+				     config->sig.sda_out,
+				     config->sig.sda_in);
+	if (r < 0) {
+		return r;
+	}
+
+	i2c_esp32_set_device_enabled(config, true);
+
+	v = *config->regs.ctr;
+
+	/* MSB or LSB first is configurable for both TX and RX */
+	if (config->mode.tx_lsb_first) {
+		v |= I2C_TX_LSB_FIRST;
+	} else {
+		v &= ~I2C_TX_LSB_FIRST;
+	}
+
+	if (config->mode.rx_lsb_first) {
+		v |= I2C_RX_LSB_FIRST;
+	} else {
+		v &= ~I2C_RX_LSB_FIRST;
+	}
+
+	if (dev_config.bits.is_master_device) {
+		/* Act as master. */
+		v |= I2C_MS_MODE;
+
+		*config->regs.slave_addr = 0;
+	} else {
+		u32_t addr;
+
+		/* Act as slave. */
+		v &= ~I2C_MS_MODE;
+
+		addr = data->address & I2C_SLAVE_ADDR_V;
+		*config->regs.slave_addr |= addr << I2C_SLAVE_ADDR_S;
+		if (dev_config.bits.use_10_bit_addr) {
+			*config->regs.slave_addr |= I2C_ADDR_10BIT_EN;
+		}
+
+		/* Before setting up FIFO and interrupts, stop transmission */
+		*config->regs.ctr &= ~I2C_TRANS_START;
+
+		/* Byte after address isn't the offset address in slave RAM */
```

Comment:
You mean "is", not "isn't", right?  Or... if not that, what is it?  Confusing comment, though maybe it makes sense if I were to read the docs.

Reply:
It's correct, but maybe this would benefit from better wording. The documentation has this as: 

> **I2C_FIFO_ADDR_CFG_EN** When this bit is set to 1, the byte received after the I2C address byte
> represents the offset address in the I2C Slave RAM